src/voip/ScreenCastPortal.cpp

```diff
@@ -0,0 +1,475 @@
+                                              QStringLiteral("Get"));
+    msg << QStringLiteral("org.freedesktop.portal.ScreenCast")
+        << QStringLiteral("AvailableSourceTypes");
+
+    QDBusPendingCall pendingCall     = QDBusConnection::sessionBus().asyncCall(msg);
+    QDBusPendingCallWatcher *watcher = new QDBusPendingCallWatcher(pendingCall);
+    connect(
+      watcher, &QDBusPendingCallWatcher::finished, this, [this](QDBusPendingCallWatcher *self) {
+          QDBusPendingReply<QDBusVariant> reply = *self;
+          self->deleteLater();
+
+          if (!reply.isValid()) {
+              nhlog::ui()->error("org.freedesktop.DBus.Properties (Get AvailableSourceTypes): {}",
+                                 reply.error().message().toStdString());
+              close();
+              return;
+          }
+
+          switch (state) {
+          case State::Closed:
+              nhlog::ui()->warn("ScreenCastPortal not starting");
+              break;
+          case State::Starting: {
+              const auto &value = reply.value().variant();
+              if (value.canConvert<uint>()) {
+                  availableSourceTypes = value.value<uint>();
+              } else {
+                  nhlog::ui()->error("Invalid reply from org.freedesktop.DBus.Properties (Get "
+                                     "AvailableSourceTypes)");
+                  close();
+                  return;
+              }
+
+              getAvailableCursorModes();
+          } break;
+          case State::Started:
+              nhlog::ui()->warn("ScreenCastPortal already started");
+              break;
+          case State::Closing:
+              break;
+          }
+      });
+}
+
+void
+ScreenCastPortal::getAvailableCursorModes()
+{
+    auto msg = QDBusMessage::createMethodCall(QStringLiteral("org.freedesktop.portal.Desktop"),
+                                              QStringLiteral("/org/freedesktop/portal/desktop"),
+                                              QStringLiteral("org.freedesktop.DBus.Properties"),
+                                              QStringLiteral("Get"));
+    msg << QStringLiteral("org.freedesktop.portal.ScreenCast")
+        << QStringLiteral("AvailableCursorModes");
+
+    QDBusPendingCall pendingCall     = QDBusConnection::sessionBus().asyncCall(msg);
+    QDBusPendingCallWatcher *watcher = new QDBusPendingCallWatcher(pendingCall);
+    connect(
+      watcher, &QDBusPendingCallWatcher::finished, this, [this](QDBusPendingCallWatcher *self) {
+          QDBusPendingReply<QDBusVariant> reply = *self;
+          self->deleteLater();
+
+          if (!reply.isValid()) {
+              nhlog::ui()->error("org.freedesktop.DBus.Properties (Get AvailableCursorModes): {}",
+                                 reply.error().message().toStdString());
+              close();
+              return;
+          }
+
+          switch (state) {
```

Comment:
You have this same block of code everywhere, but couldn't you replace that with just a `if (state == Starting) {} else { log warning }`?

Reply:
Ok, it was for separate error messages, but that's not so useful.